File: akka-sample-persistence-dc-java/src/main/java/sample/persistence/multidc/ThumbsUpHttp.java
package sample.persistence.multidc;

import akka.NotUsed;
import akka.actor.ActorRef;
import akka.actor.ActorSystem;
import akka.http.javadsl.ConnectHttp;
import akka.http.javadsl.Http;
import akka.http.javadsl.ServerBinding;
import akka.http.javadsl.model.ContentTypes;
import akka.http.javadsl.model.HttpEntities;
import akka.http.javadsl.model.HttpRequest;
import akka.http.javadsl.model.HttpResponse;
import akka.http.javadsl.server.AllDirectives;
import akka.http.javadsl.server.Route;
import akka.stream.ActorMaterializer;
import akka.stream.javadsl.Flow;
import akka.stream.javadsl.Source;
import akka.util.ByteString;
import akka.util.Timeout;

import java.util.concurrent.CompletionStage;
import java.util.concurrent.TimeUnit;

import static akka.http.javadsl.server.PathMatchers.segment;
import static akka.http.javadsl.server.PathMatchers.segments;
import static akka.pattern.PatternsCS.ask;

public class ThumbsUpHttp extends AllDirectives {

  public static void startServer(ActorSystem system, String httpHost, int httpPort, ActorRef counterRegion) {

    final ActorMaterializer materializer = ActorMaterializer.create(system);

    ThumbsUpHttp api = new ThumbsUpHttp();

    final Flow<HttpRequest, HttpResponse, NotUsed> routeFlow =
        api.createRoute(counterRegion).flow(system, materializer);
    final CompletionStage<ServerBinding> binding = Http.get(system).bindAndHandle(routeFlow,
        ConnectHttp.toHost(httpHost, httpPort), materializer);

    binding.thenAccept(b ->
        system.log().info("HTTP Server bound to http://{}:{}", httpHost, httpPort)
    );
  }

  private Route createRoute(ActorRef counterRegion) {
    Timeout timeout = Timeout.apply(10, TimeUnit.SECONDS);
    return
        pathPrefix("thumbs-up", () ->
            route(
                // example: curl http://0.0.0.0:22551/thumbs-up/a
                get(() -> {
                  return path(segment(), resourceId -> {
                    return onComplete(ask(counterRegion, new ThumbsUpCounter.GetUsers(resourceId), timeout), state -> {
                      Source<ByteString, NotUsed> s =
                          Source.fromIterator(() -> ((ThumbsUpCounter.State) state.get()).users.iterator())
                              .intersperse("\n")
                              .map(line -> ByteString.fromString(line));
                      return complete(HttpEntities.create(ContentTypes.TEXT_PLAIN_UTF8, s));
                    });
                  });
                }),
                // example: curl -X POST http://0.0.0.0:22551/thumbs-up/a/u1
                post(() -> {
                  return path(segments(2), seg -> {
                    final String resourceId = seg.get(0);
                    final String userId = seg.get(1);
                    return onComplete(ask(counterRegion, new ThumbsUpCounter.GiveThumbsUp(resourceId, userId), timeout), cnt -> {
                      return complete(cnt.get().toString());
                    });
                  });
                })
            )
        );

  }

}

File: akka-sample-persistence-dc-scala/src/main/scala/sample/persistence/multidc/ThumbsUpHttp.scala
package sample.persistence.multidc

import akka.actor.typed.ActorSystem
import akka.cluster.sharding.typed.{ReplicatedSharding, ShardingEnvelope}
import akka.http.scaladsl.Http
import akka.http.scaladsl.model.{ContentTypes, HttpEntity, HttpResponse, StatusCodes}
import akka.persistence.typed.ReplicaId
import akka.stream.scaladsl.Source
import akka.util.{ByteString, Timeout}
import sample.persistence.multidc.ThumbsUpCounter.State

import scala.concurrent.ExecutionContext
import scala.concurrent.duration._
import scala.util.{Failure, Success}

object ThumbsUpHttp {

  def startServer(httpHost: String, httpPort: Int, selfReplica: ReplicaId, res: ReplicatedSharding[ThumbsUpCounter.Command])(implicit system: ActorSystem[_]): Unit = {

    import akka.http.scaladsl.server.Directives._

    implicit val timeout: Timeout = Timeout(10.seconds)
    implicit val ec: ExecutionContext = system.executionContext

    val api = pathPrefix("thumbs-up") {
      concat(
        // example: curl http://127.0.0.1:22551/thumbs-up/a
        get {
          path(Segment) { resourceId =>
            onComplete(res.entityRefsFor(resourceId)(selfReplica).ask[State](replyTo => ThumbsUpCounter.GetUsers(resourceId, replyTo))) {
              case Success(state) =>
                val s = Source.fromIterator(() => state.users.iterator)
                  .intersperse("\n")
                  .map(ByteString(_))
                complete(HttpResponse(entity = HttpEntity(ContentTypes.`text/plain(UTF-8)`, s)))
              case Failure(ex) => complete(StatusCodes.InternalServerError, ex.toString)
            }
          }
        },
        // example: curl -X POST http://127.0.0.1:22551/thumbs-up/a/u1
        post {
          path(Segment / Segment) { (resourceId, userId) =>
            onComplete(res.entityRefsFor(resourceId)(selfReplica).ask[Long](replyTo => ThumbsUpCounter.GiveThumbsUp(resourceId, userId, replyTo))) {
              case Success(i) => complete(i.toString)
              case Failure(ex) => complete(StatusCodes.BadRequest, ex.toString)
            }
          }
        }
      )
    }

    Http().newServerAt(httpHost, httpPort).bind(api).onComplete {
      case Success(_) => system.log.info("HTTP Server bound to http://{}:{}", httpHost, httpPort)
      case Failure(ex) => system.log.error(s"Failed to bind HTTP Server to http://$httpHost:$httpPort", ex)
    }

  }
}
